v2/tools/asoctl/internal/importing/resource_importer.go

```diff
@@ -36,91 +38,157 @@ func NewResourceImporter(
 	return &ResourceImporter{
 		scheme:    scheme,
 		client:    client,
+		pending:   make(map[string]ImportableResource),
 		completed: make(map[string]ImportableResource),
 	}
 }
 
 // Add adds an importer to the list of resources to import.
 func (ri *ResourceImporter) Add(importer ImportableResource) {
-	ri.pending = append(ri.pending, importer)
+	// Lock while we're modifying the maps
+	ri.lock.Lock()
+	defer ri.lock.Unlock()
+
+	ri.addCore(importer)
 }
 
 // AddARMID adds an ARM ID to the list of resources to import.
-func (ri *ResourceImporter) AddARMID(armID string) {
-	importer := NewImportableARMResource(armID, nil /* no owner */, ri.client, ri.scheme)
+func (ri *ResourceImporter) AddARMID(armID string) error {
+	importer, err := NewImportableARMResource(armID, nil /* no owner */, ri.client, ri.scheme)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create importer for %q", armID)
+	}
+
 	ri.Add(importer)
+	return nil
+}
+
+// addCore is the internal implementation of Add that assumes the lock is already held
+// This allows us to call it from other methods that already have the lock
+func (ri *ResourceImporter) addCore(importer ImportableResource) {
```

Comment:
is `Core` being used here to mean: "The actual logic"?

I've seen `addImpl` or `addLocked` too - `core` was a bit confusing to me for about 5s as I was thinking this somehow had something to do with threading (mentioned in your PR description) and the core you were referring to was somehow a compute core.

I'm not necessarily opposed to `Core` as a suffix here it's just I haven't seen it used this way before. I do _vaguely_ think I saw in the actual Go source code usage of the `xLocked` approach, but I can't recall where so might be hallucinating it